scenes/test_string.txt

```diff
@@ -103,6 +103,28 @@
 *gosub test_finish result_expected cslib_ret
 
 
+*comment :::::: CONSISTS_OF tests ::::::
+*gosub test_start "CONSISTS_OF #1 (numbers)"
+*gosub_scene cslib_string consists_of "14756598573756237350" "0123456789"
+*temp result_expected true
+*gosub test_finish result_expected cslib_ret
+
+*gosub test_start "CONSISTS_OF #2 (spaces)"
+*gosub_scene cslib_string consists_of "       " " "
+*temp result_expected true
+*gosub test_finish result_expected cslib_ret
+
+*gosub test_start "CONSISTS_OF #3 (neg)"
+*gosub_scene cslib_string consists_of "       cslib" " "
+*temp result_expected false
+*gosub test_finish result_expected cslib_ret
+
+*gosub test_start "CONSISTS_OF #1 (single)"
+*gosub_scene cslib_string consists_of "!" "@!$%^"
```

Comment:
have also a few tests on the char set (empty, one char, repeated chars etc)

Reply:
Unfortunately you can't test an empty char set, as that'll throw a *bug, which I haven't yet worked out how to 'catch'.

Reply:
There is not "catch", you are right, but I have an idea for it, which may work. In startup.txt, we use *script to inject our modified Scene.prototype.bug. The original is below. We remove the "throw" and instead update an error variable (this.stats.cs_err = "Bug in function.."). The test can then check the value of the error message.

The library would still be *script free, but our testing framework could 'catch' the bugs.

I didn't test with randomtest or quicktest - there may be other details to fix.


---- BEFORE

Scene.prototype.bug = function scene_bug(message) {
  if (message) {
    message = "Bug: " + this.replaceVariables(message);
  } else {
    message = "Bug";
  }
  throw new Error(this.lineMsg() + message);
};

-- AFTER

*script   (everything on one line)
Scene.prototype.bug = function scene_bug(message) {
  if (message) {
    message = "Bug: " + this.replaceVariables(message);
  } else {
    message = "Bug";
  }
  this.stats.cs_err = message;
};

Reply:
I like this, but I'd prefer if we stored it as a function in a mygame.js (and made the CI consume that). Long one-line script statements are just nasty.

Reply:
Either way. If we mod *fake_choice in a similar way, we could also test cslib_menu. *fake_choice could quietly update a stat with the number of entries displayed, or even a string with all choices joined.

Reply:
Tracking the bug variant here: https://github.com/ChoicescriptIDE/cslib/issues/19

I'm less sold on making a habit of this for all commands. I override a number of stuff in CSIDE and it's honestly more of a pain than it's worth. Bug is quite low-risk as it's a simple and uncommon command that's unlikely to change. Both choice commands are quite integral and are at high risk of change with future ChoiceScript updates.

If we _really_ do want to go down that route heavily, I suggest forking ChoiceScript and making tweaks to accommodate this (and merge from upstream periodically). But a habit of blindly overriding the prototypes completely will come back to bite us.